File: leetcode/p00287/p00287.py
from typing import List

class Solution:
    # Time Complexity: O(n)
    # Space Complexity: O(1)
    def findDuplicate(self, nums: List[int]) -> int:
        slow, fast = nums[0], nums[nums[0]]
        
        while slow != fast:
            slow = nums[slow]
            fast = nums[nums[fast]]
        
        anchor = 0
        while slow != anchor:
            slow = nums[slow]
            anchor = nums[anchor]
        
        return anchor

if __name__ == "__main__":
    s = Solution()

    assert s.findDuplicate([1,3,4,2,2]) == 2
    assert s.findDuplicate([3,1,3,4,2]) == 3

File: leetcode/p00287/p00287.go
package p00287

// Time Complexity: O(nlogn)
// Space Complexity: O(1)
func findDuplicate_binarysearch(nums []int) int {
	low, high := 1, len(nums)-1

	for low < high {
		mid := low + (high-low)/2
		count := 0
		for _, num := range nums {
			if num <= mid {
				count++
			}
		}

		if count > mid {
			high = mid
		} else {
			low = mid + 1
		}
	}

	return low
}

// Time Complexity: O(n)
// Space Complexity: O(1)
func findDuplicate_cycledetection(nums []int) int {
	slow := nums[0]
	fast := nums[nums[0]]

	for slow != fast {
		slow = nums[slow]
		fast = nums[nums[fast]]
	}

	anchor := 0
	for slow != anchor {
		slow = nums[slow]
		anchor = nums[anchor]
	}

	return anchor
}
